ansi/halfblocks/encoder.go

```diff
@@ -0,0 +1,626 @@
+			// Calculate luminance
+			luma := rgbaToLuminance(block.Pixels[y][x])
+			pixelMask[y][x] = luma >= r.Options.Threshold
+		}
+	}
+
+	// Find the best matching block character
+	bestChar := ' '
+	bestScore := math.MaxFloat64
+
+	for _, blockChar := range r.Blocks {
+		score := 0.0
+		for i := 0; i < 4; i++ {
+			y, x := i/2, i%2
+			if blockChar.Coverage[i] != pixelMask[y][x] {
+				score += 1.0
+			}
+		}
+
+		if score < bestScore {
+			bestScore = score
+			bestChar = blockChar.Char
+		}
+	}
+
+	// Determine foreground and background colors based on the best character
+	var fgPixels, bgPixels []color.RGBA
+
+	// Get the coverage pattern for the selected character
+	var coverage []bool
+	for _, b := range r.Blocks {
+		if b.Char == bestChar {
+			coverage = b.Coverage
+			break
+		}
+	}
+
+	// Assign pixels to foreground or background based on the character's coverage
+	for i := 0; i < 4; i++ {
+		y, x := i/2, i%2
+		if coverage[i] {
+			fgPixels = append(fgPixels, block.Pixels[y][x])
+		} else {
+			bgPixels = append(bgPixels, block.Pixels[y][x])
+		}
+	}
+
+	// Calculate average colors
+	if len(fgPixels) > 0 {
+		block.BestFgColor = r.averageColors(fgPixels)
+	} else {
+		// Default to black if no foreground pixels
+		block.BestFgColor = color.RGBA{0, 0, 0, 255}
+	}
+
+	if len(bgPixels) > 0 {
+		block.BestBgColor = r.averageColors(bgPixels)
+	} else {
+		// Default to black if no background pixels
+		block.BestBgColor = color.RGBA{0, 0, 0, 255}
+	}
+
+	block.BestSymbol = bestChar
+}
+
+// averageColors calculates the average color from a slice of colors
+func (r *Renderer) averageColors(colors []color.RGBA) color.RGBA {
+	if len(colors) == 0 {
+		return color.RGBA{0, 0, 0, 255}
+	}
+
+	var sumR, sumG, sumB, sumA uint32
+
+	for _, c := range colors {
+		sumR += uint32(c.R)
+		sumG += uint32(c.G)
+		sumB += uint32(c.B)
+		sumA += uint32(c.A)
+	}
+
+	count := uint32(len(colors))
+	return color.RGBA{
+		R: uint8(sumR / count),
+		G: uint8(sumG / count),
+		B: uint8(sumB / count),
+		A: uint8(sumA / count),
+	}
+}
+
+// formatSymbol formats a symbol with ANSI color codes
+func (r *Renderer) formatSymbol(char rune, fg, bg color.RGBA) string {
+	if r.Options.ColorMode == 0 {
+		// No color, just return the character
+		return string(char)
+	}
+
+	var fgStr, bgStr string
+
+	switch r.Options.ColorMode {
+	case 1: // 8 colors
+		fgCode := nearestAnsi8Color(fg.R, fg.G, fg.B)
+		bgCode := nearestAnsi8Color(bg.R, bg.G, bg.B)
+		fgStr = fmt.Sprintf("\033[%dm", 30+fgCode)
+		bgStr = fmt.Sprintf("\033[%dm", 40+bgCode)
+
+	case 2: // 256 colors
+		fgCode := nearestAnsi256Color(fg.R, fg.G, fg.B)
+		bgCode := nearestAnsi256Color(bg.R, bg.G, bg.B)
+		fgStr = fmt.Sprintf("\033[38;5;%dm", fgCode)
+		bgStr = fmt.Sprintf("\033[48;5;%dm", bgCode)
+
+	case 3: // True color
+		fgStr = fmt.Sprintf("\033[38;2;%d;%d;%dm", fg.R, fg.G, fg.B)
+		bgStr = fmt.Sprintf("\033[48;2;%d;%d;%dm", bg.R, bg.G, bg.B)
+	}
+
+	return fgStr + bgStr + string(char) + "\033[0m"
```

Comment:
We should use `ansi.Style` with `color.Color` instead. Downgrading colors can should be handled by colorprofile. We have types for 16-bit and indexed in `ansi`, specifically `ansi.BasicColor`, `ansi.IndexedColor`, any other `color.Color` type will be treated as a true color.

The only optimization I can think of is checking if the RGB value corresponds to a 256 terminal color and using that to encode the color saving some SGR bytes i.e. if the color is `#5f0000` we can use the defined XTerm dark red (indexed 52) color instead of to save bytes `\x1b[48;2;95;0;0m` vs `\x1b[48;5;52m'.

Anyway, we should use `ansi.Style{}.ForegroundColor(fg).BackgroundColor(bg).Styled(string(char))` instead